src/sinks/aws_kinesis_streams/sink.rs

```diff
@@ -0,0 +1,106 @@
+use crate::buffers::Acker;
+use crate::event::{Event, LogEvent};
+use crate::sinks::aws_kinesis_streams::request_builder::{KinesisRequest, KinesisRequestBuilder};
+use crate::sinks::aws_kinesis_streams::service::KinesisResponse;
+use crate::Error;
+use futures::stream::BoxStream;
+use futures::StreamExt;
+use rand::random;
+use std::num::NonZeroUsize;
+use tower::util::BoxService;
+use vector_core::partition::NullPartitioner;
+use vector_core::stream::BatcherSettings;
+
+use crate::sinks::util::processed_event::ProcessedEvent;
+use crate::sinks::util::{SinkBuilderExt, StreamSink};
+use async_trait::async_trait;
+use futures::future;
+
+pub type KinesisProcessedEvent = ProcessedEvent<LogEvent, KinesisMetadata>;
+
+pub struct KinesisMetadata {
+    pub partition_key: String,
+}
+
+pub struct KinesisSink {
+    pub batch_settings: BatcherSettings,
+    pub acker: Acker,
+    pub service: BoxService<Vec<KinesisRequest>, KinesisResponse, Error>,
+    pub request_builder: KinesisRequestBuilder,
+    pub partition_key_field: Option<String>,
+}
+
+impl KinesisSink {
+    async fn run(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        let request_builder_concurrency_limit = NonZeroUsize::new(50);
+
+        let partition_key_field = self.partition_key_field.clone();
+        let sink = input
+            .map(|event| {
+                // Panic: This sink only accepts Logs, so this should never panic
+                event.into_log()
+            })
+            .filter_map(move |log| future::ready(process_log(log, &partition_key_field)))
+            .request_builder(request_builder_concurrency_limit, self.request_builder)
+            .filter_map(|request| async move {
+                match request {
+                    Err(e) => {
+                        error!("Failed to build Kinesis Stream request: {:?}.", e);
+                        None
+                    }
+                    Ok(req) => Some(req),
+                }
+            })
+            .batched(NullPartitioner::new(), self.batch_settings)
+            .map(|(_, batch)| batch)
+            .into_driver(self.service, self.acker);
+
+        sink.run().await
+    }
+}
+
+#[async_trait]
+impl StreamSink for KinesisSink {
+    async fn run(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        self.run(input).await
+    }
```

Comment:
Hmm, I'm surprised this works and isn't just an infinite loop.

Reply:
it does work... I'll change it anyway because it makes me uncomfortable now that you mention it